qabelbox/src/main/java/de/qabel/qabelbox/chat/interactor/TransformingChatUseCase.kt

```diff
@@ -31,13 +33,37 @@ class TransformingChatUseCase @Inject constructor(val identity: Identity, overri
     }
 
     override fun retrieve() = observable<ChatMessage> { subscriber ->
-        thread {
+        chatDropMessageRepository.markAsRead(contact, identity)
+        chatDropMessageRepository.findByContact(contact.id, identity.id).forEach { msg ->
+            subscriber.onNext(chatMessageTransformer.transform(msg))
+        }
+        subscriber.onCompleted()
+    }
+
+    override fun load(offset: Int, pageSize: Int) = observable<PagingResult<ChatMessage>> { subscriber ->
+        if(offset == 0){
             chatDropMessageRepository.markAsRead(contact, identity)
-            chatDropMessageRepository.findByContact(contact.id, identity.id).forEach { msg ->
-                subscriber.onNext(chatMessageTransformer.transform(msg))
-            }
-            subscriber.onCompleted()
         }
+        chatDropMessageRepository.findByContact(contact.id, identity.id, offset, pageSize).let { pagingResult ->
+            subscriber.onNext(pagingResult.transform { chatMessageTransformer.transform(it) })
+        }
+        subscriber.onCompleted()
     }
+
+    fun <X, T> PagingResult<T>.transform(transformer: (T) -> X): PagingResult<X> =
+            PagingResult(availableRange, result.map { transformer(it) })
+
+    override fun addContact() = observable<Unit> { subscriber ->
+        chatServiceUseCase.addContact(identity.keyIdentifier, contact.keyIdentifier)
+        subscriber.onNext(Unit)
+        subscriber.onCompleted()
+    }
+
+    override fun ignoreContact() = observable<Unit> { subscriber ->
```

Comment:
This is a single-observable from Unit:

https://github.com/ReactiveX/RxKotlin/blob/0.x/src/main/kotlin/rx/lang/kotlin/single.kt